crates/core/tedge_actors/src/servers/message_boxes.rs

```diff
@@ -123,3 +145,165 @@ impl<Request: Message, Response: Message> ClientMessageBox<Request, Response> {
             .ok_or(ChannelError::ReceiveError())
     }
 }
+
+#[cfg(test)]
+#[cfg(feature = "test-helpers")]
+mod tests {
+    use super::*;
+
+    use crate::test_helpers::ServiceProviderExt;
+    use crate::ConcurrentServerActor;
+    use crate::DynSender;
+    use crate::Runtime;
+    use crate::RuntimeRequest;
+    use crate::RuntimeRequestSink;
+    use crate::Server;
+    use crate::ServerMessageBoxBuilder;
+    use async_trait::async_trait;
+    use std::sync::Arc;
+    use std::time::Duration;
+    use tokio::sync::mpsc::error::TryRecvError;
+    use tokio::time::timeout;
+
+    #[tokio::test]
+    async fn only_processes_messages_up_to_max_concurrency() {
+        let mut builder = SimpleMessageBoxBuilder::new("ConcurrentServerMessageBoxTest", 16);
+        let mut test_box = builder.new_client_box(NoConfig);
+        let message_box: ServerMessageBox<i32, i32> = builder.build();
+        let mut concurrent_box = ConcurrentServerMessageBox::new(4, message_box);
+
+        // to pause initial 4 tasks
+        let (broadcast, _rx) = tokio::sync::broadcast::channel::<()>(16);
+
+        // use other channel to return results from tasks because it has
+        // `try_recv` which doesn't block
+        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+
+        // send all messages to the concurrent message box
+        for i in 0..5 {
+            test_box.send((i as usize, i)).await.unwrap();
+        }
+
+        let broadcast = Arc::new(broadcast);
+        let _broadcast = broadcast.clone();
+        tokio::spawn(async move {
+            loop {
+                let _tx = tx.clone();
+                let mut rx = _broadcast.subscribe();
+                let request = concurrent_box.recv().await.unwrap();
+                concurrent_box.send_response_once_done(tokio::spawn(async move {
+                    _tx.send(request).unwrap();
+                    rx.recv().await.unwrap();
+                    request
+                }));
+            }
+        });
+
+        // Expect first 4 tasks to be in-progress
+        assert_eq!(rx.recv().await, Some((0usize, 0)));
+        assert_eq!(rx.recv().await, Some((1usize, 1)));
+        assert_eq!(rx.recv().await, Some((2usize, 2)));
+        assert_eq!(rx.recv().await, Some((3usize, 3)));
+
+        // Expect at this point in time that 5th task hasn't started executing
+        // yet
+        // TODO: remove waits
```

Comment:
Why do you think these waits are useful in the current setting?

Reply:
The original attempt was to trade some of the perceived flakiness for delays. I was being overly cautious with assuming that spawned tasks might not begin executing immediately. I shuffled some channels, added some more asserts, and removed the waits, and now I'm reasonably confident that they shouldn't be flaky.